storage/src/path/mod.rs

```diff
@@ -2,5 +2,192 @@
+    /// The input paths are consumed and ownership is taken.
+    fn join_path<S>(self, suffix: S) -> JoinedPath<Self, S>
+    where
+        Self: Sized,
+        S: TriePath,
+    {
+        JoinedPath::new(self, suffix)
+    }
+
+    /// The inverse of [`TriePath::join_path`]; joins the given prefix with this path.
+    fn join_prefix<P>(self, prefix: P) -> JoinedPath<P, Self>
+    where
+        Self: Sized,
+        P: TriePath,
+    {
+        prefix.join_path(self)
+    }
+
+    /// Compares this path against another path for equality using path component
+    /// equality.
+    fn path_eq<T: TriePath>(&self, other: &T) -> bool {
+        self.len() == other.len() && self.components().eq(other.components())
+    }
+
+    /// Compares this path against another path using path-component lexicographic
+    /// ordering. Strict prefixes are less than their longer counterparts.
+    fn path_cmp<T: TriePath>(&self, other: &T) -> core::cmp::Ordering {
+        self.components().cmp(other.components())
+    }
+
+    /// Returns a wrapper type that implements [`std::fmt::Display`] and
+    /// [`std::fmt::Debug`] for this path.
+    fn display(&self) -> DisplayPath<'_, Self> {
+        DisplayPath { path: self }
+    }
+}
+
+/// Constructor for a trie path from a set of unpacked bytes; where each byte
+/// is a whole path component regardless of the normal width of a path component.
+///
+/// For 256-ary tries, this is the bytes as-is.
+///
+/// For hexary tries, each byte must occupy only the lower 4 bits. Any byte with
+/// a bit set in the upper 4 bits will result in an error.
+pub trait TriePathFromUnpackedBytes<'input>: TriePath + Sized {
+    /// The error type returned if the bytes are invalid.
+    type Error;
+
+    /// Constructs a path from the given unpacked bytes.
+    ///
+    /// For hexary tries, each byte must be in the range 0x00 to 0x0F inclusive.
+    /// Any byte outside this range will result in an error.
+    ///
+    /// # Errors
+    ///
+    /// - The input is invalid.
+    fn path_from_unpacked_bytes(bytes: &'input [u8]) -> Result<Self, Self::Error>;
+}
+
+#[inline]
+fn display_path(
+    f: &mut std::fmt::Formatter<'_>,
+    mut comp: impl Iterator<Item = PathComponent>,
+) -> std::fmt::Result {
+    comp.try_for_each(|c| write!(f, "{c}"))
+}
+
+/// A wrapper type that implements [`std::fmt::Display`] and [`std::fmt::Debug`]
+/// for any type that implements [`TriePath`].
+pub struct DisplayPath<'a, P: TriePath + ?Sized> {
+    path: &'a P,
+}
+
+impl<P: TriePath + ?Sized> std::fmt::Debug for DisplayPath<'_, P> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        display_path(f, self.path.components())
+    }
+}
+
+impl<P: TriePath + ?Sized> std::fmt::Display for DisplayPath<'_, P> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        display_path(f, self.path.components())
+    }
+}
+
+impl<T: TriePath + ?Sized> TriePath for &T {
```

Comment:
Isn't it possible to do something a bit more generic rather than re-implement this for each type that dereferences into a T? Something like:
```
impl<T, D> TriePath for D
where
    D: Deref<Target = T>,
    T: TriePath,
{
...
```

Then we don't need so many implementations for all things that dereference to T.

Reply:
No, this is not possible due to rust's foreign impl rules.